arch/arm/core/elf.c

```diff
@@ -11,42 +12,261 @@
+#define MEM2THM16OPCODE(x) OPCODE2THM16MEM(x)
+#define JUMP_UPPER_BOUNDARY ((int32_t)0xfe000000)
+#define JUMP_LOWER_BOUNDARY ((int32_t)0x2000000)
+#define PREL31_UPPER_BOUNDARY ((int32_t)0x40000000)
+#define PREL31_LOWER_BOUNDARY ((int32_t)-0x40000000)
+#define THM_JUMP_UPPER_BOUNDARY ((int32_t)0xff000000)
+#define THM_JUMP_LOWER_BOUNDARY ((int32_t)0x01000000)
+#define MASK_V4BX_RM_COND 0xf000000f
+#define MASK_V4BX_NOT_RM_COND 0x01a0f000
+#define MASK_BRANCH_COND GENMASK(31, 28)
+#define MASK_BRANCH_101 GENMASK(27, 25)
+#define MASK_BRANCH_L BIT(24)
+#define MASK_BRANCH_OFFSET GENMASK(23, 0)
+#define MASK_MOV_COND GENMASK(31, 28)
+#define MASK_MOV_00 GENMASK(27, 26)
+#define MASK_MOV_I BIT(25)
+#define MASK_MOV_OPCODE GENMASK(24, 21)
+#define MASK_MOV_S BIT(20)
+#define MASK_MOV_RN GENMASK(19, 16)
+#define MASK_MOV_RD GENMASK(15, 12)
+#define MASK_MOV_OPERAND2 GENMASK(11, 0)
+#define BIT_THM_BW_S 10
+#define MASK_THM_BW_11110 GENMASK(15, 11)
+#define MASK_THM_BW_S BIT(10)
+#define MASK_THM_BW_IMM10 GENMASK(9, 0)
+#define BIT_THM_BL_J1 13
+#define BIT_THM_BL_J2 11
+#define MASK_THM_BL_10 GENMASK(15, 14)
+#define MASK_THM_BL_J1 BIT(13)
+#define MASK_THM_BL_1 BIT(12)
+#define MASK_THM_BL_J2 BIT(11)
+#define MASK_THM_BL_IMM11 GENMASK(10, 0)
+#define MASK_THM_MOV_11110 GENMASK(15, 11)
+#define MASK_THM_MOV_I BIT(10)
+#define MASK_THM_MOV_100100 GENMASK(9, 4)
+#define MASK_THM_MOV_IMM4 GENMASK(3, 0)
+#define MASK_THM_MOV_0 BIT(15)
+#define MASK_THM_MOV_IMM3 GENMASK(14, 12)
+#define MASK_THM_MOV_RD GENMASK(11, 8)
+#define MASK_THM_MOV_IMM8 GENMASK(7, 0)
+#define SHIFT_PREL31_SIGN 30
+#define SHIFT_BRANCH_OFFSET 2
+#define SHIFT_JUMPS_SIGN 25
+#define SHIFT_MOV_RD 4
+#define SHIFT_MOV_RN 4
+#define SHIFT_MOVS_SIGN 15
+#define SHIFT_THM_JUMPS_SIGN 24
+#define SHIFT_THM_BW_IMM10 12
+#define SHIFT_THM_BL_J2 22
+#define SHIFT_THM_BL_J1 23
+#define SHIFT_THM_MOVS_SIGN 15
+#define SHIFT_THM_MOV_I 1
+#define SHIFT_THM_MOV_IMM3 4
+#define SHIFT_THM_MOV_IMM4 12
 
-static int32_t arm_bl_blx_decode_addend(uintptr_t opaddr)
+static inline int prel31_decode(elf_word reloc_type, uint32_t loc,
+				uint32_t sym_base_addr, const char *sym_name, int32_t *offset)
 {
-	uint16_t upper = *((uint16_t *)opaddr);
-	uint16_t lower = *(((uint16_t *)opaddr) + 1);
+	int ret;
 
-	int32_t addend = upper & ARM_BL_BLX_UPPER_S_BIT ? UINT32_MAX : 0;
+	*offset = sign_extend(*(int32_t *)loc, SHIFT_PREL31_SIGN);
+	*offset += sym_base_addr - loc;
+	if (*offset >= PREL31_UPPER_BOUNDARY || *offset < PREL31_LOWER_BOUNDARY) {
+		LOG_ERR("sym '%s': relocation out of range (%#x -> %#x)\n",
+			sym_name, loc, sym_base_addr);
+		ret = -ENOEXEC;
+	} else {
+		ret = 0;
+	}
+
+	return ret;
+}
+
+static inline void prel31_reloc(uint32_t loc, int32_t *offset)
+{
+	*(uint32_t *)loc &= BIT(31);
+	*(uint32_t *)loc |= *offset & GENMASK(30, 0);
+}
+
+static int prel31_handler(elf_word reloc_type, uint32_t loc,
+				uint32_t sym_base_addr, const char *sym_name, int32_t *offset)
```

Comment:
Offset is not used outside of the handler and could be made local to this block, passed by value to the reloc function